client/remoterepo.c

```diff
@@ -650,56 +636,44 @@ TDNFParseAndGetURLFromMetalink(
     }
 
     metalink_context = metalink_parser_context_new();
-    if(metalink_context == NULL)
+    if (!metalink_context)
     {
         dwError = ERROR_TDNF_OUT_OF_MEMORY;
         BAIL_ON_TDNF_ERROR(dwError);
     }
 
-    fd = open(pszFile, O_RDONLY);
-    if(fd == -1)
+    dwError = TDNFFileReadAllText(pszFile, &readbuf);
+    if (dwError)
     {
         dwError = errno;
         BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
     }
-    while((length = read(fd, buf, (sizeof(buf)-1))) > 0)
-    {
-        metalink_error = metalink_parse_update(metalink_context, buf, length);
-        memset(buf, 0, BUFSIZ);
-        if(metalink_error != 0)
-        {
-            if(metalink_context)
-            {
-                metalink_parser_context_delete(metalink_context);
-            }
-            fprintf(stderr, "Unable to parse metalink, ERROR: code=%d\n", metalink_error);
-            dwError = ERROR_TDNF_INVALID_PARAMETER;
-            BAIL_ON_TDNF_ERROR(dwError);
-        }
-    }
-    if(length == -1)
+
+    metalink_error = metalink_parse_update(metalink_context, readbuf, strlen(readbuf));
+    if (metalink_error)
     {
-        if(metalink_context)
-        {
-            metalink_parser_context_delete(metalink_context);
-        }
-        dwError = errno;
-        BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
+        fprintf(stderr, "Unable to parse metalink, ERROR: code=%d\n", metalink_error);
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
     }
+
     metalink_error = metalink_parse_final(metalink_context, NULL, 0, &metalink);
-    if((metalink_error != 0) || (metalink == NULL))
+    if (metalink_error || !metalink)
     {
         fprintf(stderr, "metalink_parse_final failed, ERROR: code=%d\n", metalink_error);
         dwError = ERROR_TDNF_INVALID_PARAMETER;
         BAIL_ON_TDNF_ERROR(dwError);
     }
-    if(metalink->files == NULL)
+    metalink_context = NULL;
```

Comment:
If metalink_context not set to NULL, cleanup will fail

Reply:
Does, `metalink_parse_final()` also get rid of the `metalink_context`? I'm not familiar with this API, but this looks buggy from an API standpoint as`metalink_parser_context_delete()`must be called.

Reply:
Yes, metalink_parse_final does metalink_parser_context_delete. Check man page: https://www.systutorials.com/docs/linux/man/3-metalink_parse_final/

Reply:
@cbsiddharth yes @sshedi is right.

Reply:
Ok, Thanks.